src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1619,7 +1620,8 @@ static void executeMergeAggregatedImpl(
         final,
         settings.distributed_aggregation_memory_efficient && is_remote_storage,
         settings.max_threads,
-        settings.aggregation_memory_efficient_merge_threads);
+        settings.aggregation_memory_efficient_merge_threads,
+        should_produce_results_in_order_of_bucket_number);
```

Comment:
Hi @nickitat , would MergingAggregatedStep always perform the final aggregation? If yes, then would should_produce_results_in_order_of_bucket_number always be false, because to_stage!= QueryProcessingStage::WithMergeableState? Could you kindly explain a bit? Thank you!

Reply:
Hi @nudles! There are might be several `MergingAggregatedStep`-s in a single query, because you may execute distributed aggregation over distributed tables. refer to [this](https://github.com/nickitat/ClickHouse/blob/5ae86e406df132d75f110651c42faffd54e6f31a/tests/queries/0_stateless/02343_aggregation_pipeline.reference#L102-L127) pipeline e.g. the first one returns [single](https://github.com/nickitat/ClickHouse/blob/5ae86e406df132d75f110651c42faffd54e6f31a/tests/queries/0_stateless/02343_aggregation_pipeline.reference#L113) stream of data, while the final merging returns [multiple](https://github.com/nickitat/ClickHouse/blob/5ae86e406df132d75f110651c42faffd54e6f31a/tests/queries/0_stateless/02343_aggregation_pipeline.reference#L106) streams.

Reply:
Hi @nickitat, thank you for the detailed explanation! One more question: is my following understanding correct?
If the `final` parameter of [MergingAggregatedStep (or AggregatingStep)](https://github.com/nickitat/ClickHouse/blob/5ae86e406df132d75f110651c42faffd54e6f31a/src/Processors/QueryPlan/MergingAggregatedStep.h#L19) is true, then to_stage is not WithMergeableState and therefore `should_produce_results_in_order_of_bucket_number=false`.

Reply:
yes, it is correct

Reply:
Hi @nickitat , then can I use `should_produce_results_in_order_of_bucket_number= !final && settings.distributed_aggregation_memory_efficient` inside `MergingAggregatedStep` and `AggregatingStep` to eliminate passing the parameter `should_produce_results_in_order_of_bucket_number` to the constructor?

Reply:
likely, yes. it worth trying not only to make code cleaner but also because we have one [place](https://github.com/ClickHouse/ClickHouse/blob/193cd1b3b2abcc13857f8660aa6995cac00b47c7/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L395) where we potentially build pipeline sub-optimally. maybe we will be able to remove [this](https://github.com/ClickHouse/ClickHouse/blob/193cd1b3b2abcc13857f8660aa6995cac00b47c7/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L457) unconditional resize.
we would be grateful if you make a pr.